rasa/core/channels/twilio_voice.py

```diff
@@ -0,0 +1,116 @@
+from sanic.request import Request
+from sanic.response import HTTPResponse
+from twilio.twiml.voice_response import VoiceResponse, Gather
+from typing import Text, Callable, Awaitable, List
+
+from rasa.shared.core.events import BotUttered
+from rasa.core.channels.channel import (
+    InputChannel,
+    CollectingOutputChannel,
+    UserMessage,
+)
+
+
+class TwilioVoiceInput(InputChannel):
+    @classmethod
+    def name(cls) -> Text:
+        """Name of your custom channel."""
+        return "twilio_voice"
+
+    def blueprint(
+        self, on_new_message: Callable[[UserMessage], Awaitable[None]]
+    ) -> Blueprint:
+
+        twilio_voice_webhook = Blueprint(
+            "custom_webhook_{}".format(type(self).__name__),
+            inspect.getmodule(self).__name__,
+        )
+
+        @twilio_voice_webhook.route("/", methods=["GET"])
+        async def health(request: Request) -> HTTPResponse:
+            return response.json({"status": "ok"})
+
+        @twilio_voice_webhook.route("/webhook", methods=["POST"])
+        async def receive(request: Request):
+            sender_id = request.form.get("From")  # method to get sender_id
+            text = request.form.get("SpeechResult")  # method to fetch text
+            input_channel = self.name()  # method to fetch input channel
+            call_status = request.form.get("CallStatus")  # method to fetch call status
+            call_sid = request.form.get("CallSid")  # Get the call identifier
+
+            collector = TwilioVoiceCollectingOutputChannel()
+
+            # Provide an initial greeting to answer the user's call.
+            if (text is None) & (call_status == "ringing"):
+                text = "hello"
+
+            # determine the response.
+            if text is not None:
+                await on_new_message(
+                    UserMessage(
+                        text, collector, sender_id, input_channel=input_channel,
+                    )
+                )
+
+                # Parse the text responses and build the Twilio VoiceResponse.
+                respond_segments = []
+                for message in collector.messages:
+                    respond_segments.append(message["text"])
+                    if "buttons" in message:
+                        for button in message["buttons"]:
+                            respond_segments.append(button["title"])
+                twilio_response = build_twilio_voice_response(respond_segments)
+                return response.text(str(twilio_response), content_type="text/xml")
+            # If the user doesn't respond to the previous message resend the last message.
+            elif text is None:
+                # Get last user utterance from tracker.
+                tracker = request.app.agent.tracker_store.retrieve(sender_id)
+                last_response = next(
+                    (e for e in reversed(tracker.events) if isinstance(e, BotUttered)),
+                    None,
+                )
```

Comment:
why do we need this?

Reply:
This is a voice channel nuance. If there is a period of silence the bot will repeat the last utterance so the user knows they need to provide a response and remind them of what they need to respond to. Happy to implement this another way, but this seemed like the easiest way to get the last utterance to repeat.

Reply:
@wochinge is there a default action for repeating the last utterance or getting the last user utterance? I didn't want to require users to have to create a custom action to implement this channel. This was the best workaround I could find if there is another approach I'm open to that.

Reply:
we could do a new default intent + default action for that 🤔 

Let's keep it like it is for now for the sake of doing a quick iteration. In the end it's just a in a channel component and not in the main architecture. Consider double checking maybe if an agent is available.